lib/membrane_opus/parser.ex

```diff
@@ -1,79 +1,375 @@
-  Parses self-delimiting Opus stream.
+  Parses a raw incoming Opus stream and adds caps information, as well as metadata.
 
-  The self-delimiting version of Opus packet contains all the information about frame sizes
-  needed to separate the packet from byte stream and decode it. This parser drops the last frame size
-  from each packet header, converting packets to the basic version. The remaining frame size needs
-  to be derived from entire packet size.
-  See https://tools.ietf.org/html/rfc6716#appendix-B for details.
+  Adds the following metadata:
+
+  duration :: non_neg_integer()
+    Number of nanoseconds encoded in this packet
   """
+
   use Membrane.Filter
 
+  import Membrane.Time
+
   alias Membrane.{Buffer, Opus, RemoteStream}
-  alias Membrane.Opus.PacketUtils
+
+  def_options delimitation: [
+                spec: :delimit | :undelimit | :keep,
+                default: :keep,
+                description: """
+                If input is delimitted (as indicated by the `self_delimiting?`
+                field in %Opus) and `:undelimit` is selected, will remove delimitting.
+
+                If input is not delimitted and `:delimit` is selected, will add delimitting.
+
+                If `:keep` is selected, will not change delimiting.
+
+                Otherwise will raise.
+
+                See https://tools.ietf.org/html/rfc6716#appendix-B for details
+                on the self-delimitting Opus format.
+                """
+              ],
+              input_delimitted?: [
+                spec: boolean(),
+                default: false,
+                description: """
+                If you know that the input is self-delimitted but you're reading from
+                some element that isn't sending the correct structure, you can set this
+                to true to force the Parser to assume the input is self-delimitted.
+                """
+              ]
 
   def_input_pad :input,
     demand_unit: :buffers,
-    caps: [{Opus, self_delimiting?: true}, {RemoteStream, content_format: one_of([Opus, nil])}]
+    caps: [
+      Opus,
+      {RemoteStream, type: :packetized, content_format: one_of([Opus, nil])}
+    ]
 
-  def_output_pad :output, caps: {Opus, self_delimiting?: false}
+  def_output_pad :output, caps: Opus
 
   @impl true
-  def handle_init(_opts) do
-    {:ok, %{partial_packet: <<>>, timestamp: 0}}
+  def handle_init(%__MODULE__{} = options) do
+    {:ok, options |> Map.from_struct()}
   end
 
   @impl true
-  def handle_caps(:input, _caps, _ctx, state) do
-    {:ok, state}
+  def handle_demand(:output, bufs, :buffers, _ctx, state) do
+    {{:ok, demand: {:input, bufs}}, state}
   end
 
   @impl true
-  def handle_demand(:output, _size, _unit, _ctx, state) do
-    {{:ok, demand: {:input, 1}}, state}
+  def handle_process(:input, %Buffer{payload: data}, ctx, state) do
+    input_delimitation =
+      Map.get(ctx.pads.input, :caps) || %{} |> Map.get(:self_delimiting?, state.input_delimitted?)
```

Comment:
this expression, in particular, i'm unhappy with. i'm sure there has to be a better way to express this?

Reply:
I found that `ctx.pads.input.caps` was `nil` at a certain point in the test phase unless I did this test :(

Reply:
Eh, that's probably because the testing sink doesn't send the caps :/ We need to finally clean that up. So you may split that into
```suggestion
      caps = Map.get(ctx.pads.input, :caps, %{})
      input_delimited? = Map.get(caps, :self_delimiting?, state.input_delimitted?)
```